ci/ray_ci/tester.py

```diff
@@ -74,40 +74,36 @@ def main(
     os.chdir(bazel_workspace_dir)
     docker_login(_DOCKER_ECR_REPO.split("/")[0])
 
-    if not build_name:
-        build_name = f"{team}build"
-    container = TesterContainer(build_name)
+    container = _get_container(
+        team, workers, worker_id, parallelism_per_worker, build_name
+    )
     if run_flaky_tests:
         test_targets = _get_flaky_test_targets(team)
     else:
         test_targets = _get_test_targets(
             container,
             targets,
             team,
-            workers,
-            worker_id,
             except_tags,
         )
-    success = container.run_tests(test_targets, test_env, parallelism_per_worker)
+    success = container.run_tests(test_targets, test_env)
     sys.exit(0 if success else 1)
 
 
-def _get_test_targets(
-    container: TesterContainer,
-    targets: str,
+def _get_container(
     team: str,
     workers: int,
     worker_id: int,
-    except_tags: Optional[str] = "",
-    yaml_dir: Optional[str] = None,
-) -> List[str]:
-    """
-    Get test targets to run for a particular shard
-    """
-    return shard_tests(
-        _get_all_test_targets(container, targets, team, except_tags, yaml_dir=yaml_dir),
-        workers,
-        worker_id,
+    parallelism_per_worker: int,
+    build_name: Optional[str] = None,
+) -> TesterContainer:
+    return TesterContainer(
+        build_name or f"{team}build",
+        shard_count=workers * parallelism_per_worker,
+        shard_ids=[
+            worker_id * parallelism_per_worker + i
+            for i in range(parallelism_per_worker)
+        ],
```

Comment:
I guess this is the part that is sharding the shards?

maybe it would be clearer if the logic understands the 2 level of sharding as the very beginning, and hence generates a 2 dimension array or something? and you can have a test that against the generated 2d array?

Reply:
yes, this is the sharding logic; the sharding logic now quires 2 dimensional thinking so it's actually more complex in that way, but it is cheaper math